gcloud-java-dns/src/main/java/com/google/gcloud/testing/LocalDnsHelper.java

```diff
@@ -0,0 +1,1426 @@
+
+    private String method;
+    private String pathRegex;
+
+    CallRegex(String method, String pathRegex) {
+      this.pathRegex = pathRegex;
+      this.method = method;
+    }
+  }
+
+  /**
+   * Wraps DNS data by adding a timestamp and id which is used for paging and listing.
+   */
+  static class RrsetWrapper {
+    static final Function<ResourceRecordSet, RrsetWrapper> WRAP_FUNCTION =
+        new Function<ResourceRecordSet, RrsetWrapper>() {
+          @Nullable
+          @Override
+          public RrsetWrapper apply(@Nullable ResourceRecordSet input) {
+            return new RrsetWrapper(input);
+          }
+        };
+    private final ResourceRecordSet rrset;
+    private final Long timestamp = System.currentTimeMillis();
+    private String id;
+
+    RrsetWrapper(ResourceRecordSet rrset) {
+      // The constructor creates a copy in order to prevent side effects.
+      this.rrset = new ResourceRecordSet();
+      this.rrset.setName(rrset.getName());
+      this.rrset.setTtl(rrset.getTtl());
+      this.rrset.setRrdatas(ImmutableList.copyOf(rrset.getRrdatas()));
+      this.rrset.setType(rrset.getType());
+    }
+
+    void setId(String id) {
+      this.id = id;
+    }
+
+    String id() {
+      return id;
+    }
+
+    /**
+     * Equals does not care about the listing id and timestamp metadata, just the rrset.
+     */
+    @Override
+    public boolean equals(Object other) {
+      return (other instanceof RrsetWrapper) && Objects.equals(rrset, ((RrsetWrapper) other).rrset);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(rrset);
+    }
+
+    ResourceRecordSet rrset() {
+      return rrset;
+    }
+
+    @Override
+    public String toString() {
+      return MoreObjects.toStringHelper(this)
+          .add("rrset", rrset)
+          .add("timestamp", timestamp)
+          .add("id", id)
+          .toString();
+    }
+  }
+
+  /**
+   * Associates a project with a collection of ManagedZones. Thread safe.
+   */
+  static class ProjectContainer {
+    private final Project project;
+    private final ConcurrentSkipListMap<String, ZoneContainer> zones =
+        new ConcurrentSkipListMap<>();
+
+    ProjectContainer(Project project) {
+      this.project = project;
+    }
+
+    Project project() {
+      return project;
+    }
+
+    ConcurrentSkipListMap<String, ZoneContainer> zones() {
+      return zones;
+    }
+  }
+
+  /**
+   * Associates a zone with a collection of changes and dns records. Thread safe.
+   */
+  static class ZoneContainer {
+    private final ManagedZone zone;
+    /**
+     * DNS records are held in a map to allow for atomic replacement of record sets when applying
+     * changes. The key for the map is always the zone name. The collection of records is immutable
+     * and must always exist, i.e., dnsRecords.get(zone.getName()) is never {@code null}.
+     */
+    private final ConcurrentSkipListMap<String, ImmutableList<RrsetWrapper>>
```

Comment:
Instead of using a map with one key:value pair to force atomic gets/sets, would using the "volatile" keyword work here? /cc @aozarov